### PR TITLE
[trainer] Support parsing test results from SwiftTesting

### PR DESCRIPTION
**Pull request by AliSoftware**
This PR allows `trainer` (and by extension `scan`/`run_tests`) to support test results from [Swift Testing](https://developer.apple.com/documentation/Testing)—by parsing the `.xcresult` bundle using the new `xcresulttool` API

### ☑️  Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.
- [x] I've added or updated relevant unit tests.

### ❔  Motivation and Context

In Xcode 16b3, the `xcresulttool` tool—that ships with Xcode Command Line Tools to parse and extract data from `.xcresult` bundles—had its CLI API updated.

So far we've relied on adding the `--legacy` flag to our `xcresulttool` invocations to be able to still execute old CLI and still use the old way of calling the tool, using `xcresulttool get <object> --format json` / `xcresulttool get <object> --id <id>`. But:
 - As the name of the flag suggests, this way of calling the tool is legacy, and might become unsupported in future versions
 - That way of calling the tool seems to not support extracting information from reports from tests written using Swift Testing, and only supports tests written in XCTest.

The version of `xcresulttool` that ships with Xcode 16+ now expects us to call the CLI with `xcresulttool get test-results test …`, and generates a new JSON format (whose JSONSchema is described in `xcresulttool get test-results test --help`) that is easier to parse but more importantly also supports Swift Testing test results.

### ✍️  Description

Instead of trying to modify the implementation of `xcresult.rb` to adapt to the new command and JSON format—which would make the code difficult to read if we wanted to keep backwards compatibility in the same codebase—and because the new JSON format is way better structured and doesn't need the complexity of what the old XCResult parser implementation needed, I've decided to instead avoid piling up on the legacy implementation and implement the parsing of new JSON format completely separately.

I have thus:
 - Extracted the methods dedicated to parsing the legacy `.xcresult` format as well as the old `.plist` format in dedicated classes and files, to separate that legacy code clearly.
 - Renamed the class dedicated to parsing the legacy `.xcresult` format to `LegacyXCResult`
 - Created a new `XCResult` class instead, dedicated to parsing the new format
 - Made `TestParser` call the `parse_xcresult` method of either `LegacyXCResult` or `XCResult` class depending on the version of `xcresulttool` being used/installed (with a possibility to force the legacy mode if needed, to avoid breaking things accidentally if we missed some edge cases in the new implementation

The new implementation of `XCResult` parsing is also now split into nested classes under `xcresult/*.rb`, namely `TestPlan`, `TestSuite`, `TestCase`, and `Repetitions`, which make semantic sense and are easier to reason about when parsing the `xcresult` JSON representation.

The generation of the output JUnit file is also directly handled in the model objects (`to_xml` methods) rather than using ERB (via `JunitGenerator`) like the old implementation was using.

### 🧪  Testing Steps

I've added a couple of Unit Tests to `test_parser_spec.rb`, that use an `.xcresult` file that I generated from a dummy `xcproject`[^1], which contained a mix of Unit and UITests, implemented with a mix of XCTest and SwiftTesting, and using features like test skipping, arguments, and retries.

Hopefully this should cover a good set of features that those testing frameworks offer.

That being said, I'd love if more people would test this implementation against some `.xcresult` files of their own Xcode projects, to validate the generated JUnit `.xml` looks right and contains all necessary info (failure messages, etc) and so that it covers a whole range of possibilities (e.g. tests runs with or without retries enabled, custom traits, etc…)

### 👁️  Suggestions on how to review this PR

This PR's diff is huge, but this is mostly because:
 - A lot of code was just moved (as opposed to modified/added/deleted)
 - I included large fixtures (`.junit` expectation fixtures which are XMLs of >1K lines each)
 - I commited the dummy Xcode project I used for generating the `.xcresult` fixture, even if that Xcode project is not directly referenced by any of the rspec tests, but so that it'd be easy to add more tests in the future in we wanted to regenerate the `.xcresult` to cover more edge cases
 - ~~I also had to commit the new `.xcresult` fixture… which is in practice a bundle containing a lot of small (mostly-binary) files. So that shows up as a lot of files in this PR's diff, while those don't need to be manually reviewed at all when reviewing that PR.~~ [EDIT] I ended up only commiting the `.xcresult/database.sqlite3` file as I discovered that it was enough for `xcresulttool` to operate.

I'd thus suggest the reviewers to instead focus on:
 - Changes in `trainer/lib/*` — keeping in mind that some of those changes are just existing code for the legacy parsers having been moved to dedicated files 
 - Unit tests in `trainer/spec/*_spec.rb`
 - For `trainer/spec/fixtures`, only check the `*.junit` expectation fixtures that were added. You can in particular safely ignore:
    - The whole `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting.xcresult/*` folder and its files, since that fixture was auto-generated (it's the bundle that Xcode generated from my dummy project)
    - The intermediate `Xcode16-Mixed-XCTest-SwiftTesting.json` fixture, as that fixture was also auto-generated (by running `xcresulttool` on the added `.xcresult` bundle fixture)
    - The whole `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting-Project/` folder, since this just contains the dummy Xcode project with lots of dummy tests that was used to generate the `.xcresult` fixture. Of course it's ok if you want to review the Xcode project itself to compare how those various ways of writing tests in Swift appear in the JUnit XML fixture files, but it's probably not worth reviewing that in GitHub directly if so, but directly in Xcode instead.

Hopefully those suggestions should help reduce the scope of the review of this PR by a lot

[^1]: Created with Cursor and AI agent, so I could quickly generate a full project with a wide range of all possible API usages and edge cases of XCTest and SwiftTesting features

### CODE REVIEW

**Comment by TheMetalCode** [comment]
Happy to test this with my projects! Thank you! 👏 

I had just started to figure out Swift Testing issues and had started testing with https://github.com/fastlane/fastlane/pull/29470, which addressed issues with test cases nested in inner classes and included some tests specific to Swift Testing, but I saw no way to handle parameterized tests (see https://github.com/fastlane/fastlane/issues/29174) and a more explicit/clean way to detect and remove auto-retries without using the newer, non-legacy `xcresulttool` output and new parsing/modeling code to match. Evidently I missed the presence of this PR until today. 🤦

**Comment by TheMetalCode** [comment]
@AliSoftware One thing I'm seeing so far in my testing:

```bash
[16:41:04]: Called from Fastfile at line 154
[16:41:04]: ```
[16:41:04]:     152:      desc "Uses trainer to collect test results after a test run"
[16:41:04]:     153:      lane :collect_test_results do |options|
[16:41:04]:  => 154:        trainer(
[16:41:04]:     155:          path: "#{options[:test_plan]}_test_result_bundle/Test-#{options[:test_plan]}.xcresult",
[16:41:04]:     156:          output_directory: "test_results",
[16:41:04]: ```
[16:41:04]: Illegal character "\u0001" in raw string "audioPlayerReturnedForOnlyValidInput(\u0001)"
```

For a test that looks something like:

```Swift
@Test(.serialized, arguments: (0..<128).map(UnicodeScalar.init).map(Character.init))
func audioPlayerReturnedForOnlyValidInput(asciiCharacter: Character) throws { ... }
```

Hopefully some kind of sanitization we can do on test names? Similar to https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/legacy_xcresult.rb#L180?

**Comment by AliSoftware** [comment]
Thanks so much @TheMetalCode for testing this PR against some more concrete tests!

> Hopefully some kind of sanitization we can do on test names? Similar to https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/legacy_xcresult.rb#L180?

Ah, good catch. Yeah, that sanitization seems like a decent plan and easy solution.

I'd like to better understand the root cause of this issue first though rather than blindly applying a workaround that was applied at the time the first `trainer` implementation was done, because maybe we can come up with something better / more resilient? Like, depending on at which level in the pipeline this error happens (during the execution of `xcresulttool` itself? During Ruby's parsing of `xcresulttool`'s JSON stdout to a Hash object? Later in the process when trying to generate the XML?) maybe if we force the use of the right unicode encoding at the right level in the code to avoid the issue altogether? 🤔

**Comment by AliSoftware** [comment]
Btw, for refs, below is the dummy Xcode project I used to generate the `trainer/spec/fixtures/Xcode16-Mixed-XCTest-SwiftTesting.xcresult` bundle that I use in my fixtures.

[FastlaneTrainerExample.zip](https://github.com/user-attachments/files/19067975/FastlaneTrainerExample.zip)

Maybe I should commit it in the repo, even if the project itself is not used by the tests, so that it'd be easier to re-generate the `.xcresult` bundle with more tests if we want to add more test cases (like the one you made with tricky unicode characters)?

Speaking of which, feel free to add your test case from above to that dummy Xcode project then regenerate the `.xcresult` and `.json` fixtures with that added test, so we can make sure we cover such edge-case (via a future commit that'd fix this and take care of those kind of uncommon characters)

**Comment by TheMetalCode** [comment]
@AliSoftware Your latest push definitely resolves the issue I was seeing on my end with the ASCII chars, and looking at your sample Xcode project I think you're covering all the edge cases I could think of. I think it's a great idea to check that into the repo too - adds transparency and helps with future maintenance for sure.

The only issue I see now is that CircleCI throws a non-fatal error when uploading the xmls produced by this trainer patch to test results:

run_unit_tests-News__667182__-_doximity_iOS-Doximity

Which is....wrong? `testsuites` and `properties` are totally valid elements to use in a JUnit xml!

Comparing the xmls generated by trainer currently to those generated by this branch, I really don't see a lot of difference in terms of xml structure. For the top-level `testsuites` element I'm seeing `<testsuites tests='1037' failures='0' skipped='1' time='4923.3454'>` vs `<testsuites tests="990" failures="0">`, so more attributes at the top level. And I guess current `trainer` doesn't put `properties` under a `testcase` but again, that's perfectly valid for JUnit xml, see https://github.com/testmoapp/junitxml/blob/main/examples/conventions.xml

The only other difference I notice is single-quotes for attribute values instead of double-quotes, not sure if that's an issue or not.

And it's a non-fatal error - it doesn't fail the CI job, nor does it fail to populate the test results tab:
Cursor_and_run_unit_tests-News__667182__-_doximity_iOS-Doximity

File: trainer/lib/trainer/xcresult/test_plan.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          skipped: test_suites.sum(&:skipped_count).to_s,
+          time: test_suites.sum(&:duration).to_s)
+
+        # Add each test suite to the root
+        test_suites.each do |suite|
+          testsuites.add_element(suite.to_xml(output_remove_retry_attempts: output_remove_retry_attempts))
+        end
+
+        # Convert to XML string with prologue
+        doc = REXML::Document.new
```
Comment (TheMetalCode) [inline trainer/lib/trainer/xcresult/test_plan.rb:62]:
```suggestion
        doc = REXML::Document.new
        doc.context[:attribute_quote] = :quote
```

See https://github.com/fastlane/fastlane/pull/29463#issuecomment-2704511016

Not sure how much single vs double quote matters for attribute values but I believe this would make it double-quotes to be consistent with what [the legacy junit generator](https://github.com/fastlane/fastlane/blob/alisoftware/trainer/support-swift-testing/trainer/lib/trainer/junit_generator.rb) would produce.

Comment (TheMetalCode) [reply trainer/lib/trainer/xcresult/test_plan.rb:62]: The usage of `:attribute_quote` doesn't appear to be explicitly documented :(

But you can see it in action here: https://github.com/ruby/rexml/blob/master/test/test_attributes.rb#L181

File: trainer/lib/trainer/xcresult/test_plan.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        end
+
+        @devices.each do |device|
+          device_prop = Helper.create_xml_element('property', name: 'device', value: "#{device.fetch('modelName', 'Unknown Device')} (#{device.fetch('osVersion', 'Unknown OS Version')})")
+          properties.add_element(device_prop)
+        end
+
+        testsuites.add_element(properties) if properties.elements.any?
+
+        doc.add(testsuites)
```
Comment (TheMetalCode) [inline trainer/lib/trainer/xcresult/test_plan.rb:80]: Ok upon further inspection, I think the top-level `testsuites` element is not supposed to have `properties`, so this might actually be the source of the issue I referenced in https://github.com/fastlane/fastlane/pull/29463#issuecomment-2704511016

https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd
https://github.com/testmoapp/junitxml/blob/main/examples/junit-complete.xml

A `testsuite` element can have `properties`, as can a `testcase`, but the top-level `testsuites` element is not expected to have `properties`. 

So when CircleCI says this:
```bash
* invalid testsuites element: properties
* invalid top level end element: testsuites
```

It might be saying that `testsuites` is invalid on account of having `properties`.

It's probably going to seem tedious to put `Configuration` and `device` properties under every `testsuite` but that does appear to be what the JUnit schema expects.

Comment (TheMetalCode) [reply trainer/lib/trainer/xcresult/test_plan.rb:80]: On my end, I was able to do a REXML `delete_element("/testsuites/properties")` before the CircleCI `upload_test_results`. I can confirm that having `properties` in the top-level `testsuites` is the only issue, no errors are thrown once those are gone.

Comment (AliSoftware) [reply trainer/lib/trainer/xcresult/test_plan.rb]: Nice catch and investigation!

I've just pushed [7deb84b](https://github.com/fastlane/fastlane/pull/29463/commits/7deb84b030ed4c5e60871530e09a9573811529b0) which should fix this

File: trainer/lib/trainer/xcresult/test_plan.rb (hunk 1 above)
Comment (AliSoftware) [reply trainer/lib/trainer/xcresult/test_plan.rb:79]: Nice one!

But since those single vs double quotes were not the core of the issue CircleCI was having with the XML files (as it was instead the `<properties>` not being expected at the top-level element, as you later discovered), and with the fact that `:attribute_quote` is not officially documented, I'd prefer to avoid using it / relying on it in our implementation after all, and stick to the official approach and defaults of how REXML generates and quotes the attributes. wdyt?

Comment (TheMetalCode) [reply trainer/lib/trainer/xcresult/test_plan.rb:79]: Agreed, probably not worth the trouble for something that wasn't actually causing problems.

File: trainer/lib/trainer/xcresult/test_plan.rb (hunk 2 above)
Comment (TheMetalCode) [reply trainer/lib/trainer/xcresult/test_plan.rb]: Confirmed the fix, thank you!

File: trainer/lib/trainer/legacy_xcresult.rb [hunk 3]
```diff
@@ -0,0 +1,586 @@
+require 'open3'
+
+require_relative 'xcresult/helper'
+
+module Trainer
+  module LegacyXCResult
+    # Model attributes and relationships taken from running the following command:
+    # xcrun xcresulttool formatDescription
```
Comment (TheMetalCode) [inline trainer/lib/trainer/legacy_xcresult.rb:8]:
```suggestion
    # xcrun xcresulttool formatDescription --legacy
```

Not major since the name of the module is now `LegacyXCResult` so it's possible the `--legacy` flag would be inferred, but just in case anybody was trying to run this command and didn't make the inference.

Comment (AliSoftware) [reply trainer/lib/trainer/legacy_xcresult.rb:8]: Nice catch!

File: trainer/lib/trainer/plist_test_summary_parser.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+          summary_row[:number_of_retries] = 0
+
+          summary_row
+        end
+        data
+      end
+
+      def ensure_file_valid!(raw_json)
+        format_version = raw_json["FormatVersion"]
+        supported_versions = ["1.1", "1.2"]
```
Comment (mollyIV) [inline trainer/lib/trainer/plist_test_summary_parser.rb:53]: Does it mean that with every new version of Xcode, we need to check if the format has changed and support the new version if necessary?

File: trainer/lib/trainer/test_parser.rb [hunk 5]
```diff
@@ -102,15 +101,16 @@ def initialize(path, config = {})
       UI.user_error!("File not found at path '#{path}'") unless File.exist?(path)
 
       if File.directory?(path) && path.end_with?(".xcresult")
-        parse_xcresult(path, output_remove_retry_attempts: config[:output_remove_retry_attempts])
+        klass = XCResult::Helper.supports_xcode16_xcresulttool? && !config[:force_legacy_xcresulttool] ? XCResult::Parser : LegacyXCResult::Parser
```
Comment (mollyIV) [inline trainer/lib/trainer/test_parser.rb:104]: nitpick: 

```suggestion
        parser = XCResult::Helper.supports_xcode16_xcresulttool? && !config[:force_legacy_xcresulttool] ? XCResult::Parser : LegacyXCResult::Parser
```

File: trainer/lib/trainer/xcresult/helper.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      #
+      # @return [Boolean] Whether the xcresulttool supports Xcode 16+ commands
+      def self.supports_xcode16_xcresulttool?
+        # e.g. DEVELOPER_DIR=/Applications/Xcode_16_beta_3.app
+        # xcresulttool version 23021, format version 3.53 (current)
+        match = `xcrun xcresulttool version`.match(/xcresulttool version (?<version>[\d.]+)/)
+        version = match[:version]
+
+        Gem::Version.new(version) >= Gem::Version.new(23_021)
+      rescue
```
Comment (mollyIV) [inline trainer/lib/trainer/xcresult/helper.rb:48]: Hmm, it seems a `begin` block is missing, isn't it?

File: trainer/lib/trainer/xcresult/repetition.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
+require_relative 'test_case_attributes'
+
+module Trainer
+  module XCResult
+    class Repetition
```
Comment (mollyIV) [inline trainer/lib/trainer/xcresult/repetition.rb:5]: What does `repetition` sand for here?

File: trainer/lib/trainer/options.rb [hunk 8]
```diff
@@ -50,6 +50,11 @@ def self.available_options
                                      description: "Produces class name and test name identical to xcpretty naming in junit file",
                                      is_string: false,
                                      default_value: false),
+        FastlaneCore::ConfigItem.new(key: :force_legacy_xcresulttool,
```
Comment (mollyIV) [inline trainer/lib/trainer/options.rb:53]: This new option key is going to help to rollout this implementation with minimized regression risk 👏

**Comment by mollyIV** [comment]
Hey Ali,

Thanks a lot for this PR! Supporting SwiftTesting is a very much welcomed addition, and I truly appreciate the effort you put into it.

Also, a special shoutout for the well-written PR description. It was super informative and made reviewing this big PR much smoother. The clear steps you provided helped me a lot!

Great work, and thanks again for contributing! 🚀

**Comment by mollyIV** [comment]
> even if that project is not directly used as a fixture by the rspec tests, it's useful to have it around for the future

I agree, we could easily add more test cases in the future there 👍

File: trainer/lib/trainer/plist_test_summary_parser.rb (hunk 4 above)
Comment (AliSoftware) [reply trainer/lib/trainer/plist_test_summary_parser.rb:53]: Yes and no.

To the best of my understanding, this `.plist` format is legacy format, even older than before `.xcresult` bundles and `xcresulttool` were introduced. This is just existing code that was initially in `xcresult.rb` and that I extracted in its own file just like I did for `legacy_xcresult.rb` so we keep that backwards compatibility, but in practice I'm not sure any modern version of Xcode produces those `.plist` file formats for test summaries anymore, so I don't see a new `FormatVersion` being added ever by now.

File: trainer/lib/trainer/xcresult/helper.rb (hunk 6 above)
Comment (AliSoftware) [reply trainer/lib/trainer/xcresult/helper.rb:48]: Nope, `begin…end` is only needed when you need to specify a smaller scope for `rescue` than the current scope. Here without an explicit `begin…end`, the `rescue` applies to the whole scope of the `def …` method: there's no need to encapsulate the whole body of `def method_name … end` within a `begin…end` in those cases, as it's implicit already.

See [this rubocop rule](https://docs.rubocop.org/rubocop/cops_style.html#styleredundantbegin) for details and examples.

File: trainer/lib/trainer/xcresult/repetition.rb (hunk 7 above)
Comment (AliSoftware) [reply trainer/lib/trainer/xcresult/repetition.rb:5]: It's for when you enabled retries in your test plan Xcode (e.g. either retry on failure, or retry N times, etc).

I hesitated into naming it `class Retries` instead of `class Repetition`, but the issue with naming it `Retries` is that it suggests it doesn't count the first run, only the retries if the first run failed. While in reality if the test was run 3 times—i.e. it was run once, failed, was retried once, failed again, and was retried a 2nd time), the JSON output of `xcresulttool` will contain 3 repetitions/retries, not 2.

Besides, [the `nodeType` in the JSON returned by `xcresulttool` is named `Repetition` anyway](https://github.com/fastlane/fastlane/pull/29463/files#diff-cf9f90c4c8e47261aaaa6ccb8091fa2a22d0a0799b5043650cebff8b02c3a395R32), not `Retries`, so I figured I'd match the naming.

File: trainer/lib/trainer/test_parser.rb (hunk 5 above)
Comment (AliSoftware) [reply trainer/lib/trainer/test_parser.rb]: Fixed in 7bd6bcdd9

File: trainer/lib/trainer/xcresult/repetition.rb (hunk 7 above)
Comment (AliSoftware) [reply trainer/lib/trainer/xcresult/repetition.rb:8]: Added some more code comments in 30f32894f to provide some additional doc about this.

**Comment by AliSoftware** [comment]
@mollyIV I should have addressed all your suggestions, so ready for a re-review (and hopefully approval) 🙂

**Review by TheMetalCode — APPROVED** [review]
For whatever my approval is worth: I've been continuously running this branch on two iOS projects over the last week or so, just over 10K test cases, about 60-70% of which are Swift Testing. Aside from the ASCII character parsing and junit schema issues that have already been addressed with follow-up commits, I've seen no issues and can happily report that all failing test cases, Swift Testing or otherwise, are being properly reflected in the XML report.

This is a very welcome and deeply appreciated change, thank you! 👏

**Review by mollyIV — COMMENTED** [review]
Hi 👋 

I tested the changes on some of our projects - mix of Swift Testing and XCTests, pure XCTests and pure Swift Testing, and all were looking great! I didn't spot any issues. 

Thanks again for the changes, Ali. 

Let's ship it 🫡

**Review by mollyIV — APPROVED** [review]
